core/src/main/java/org/elasticsearch/index/shard/IndexShardOperationPermits.java

```diff
@@ -112,24 +175,25 @@ public void close() {
     /**
      * Acquires a permit whenever permit acquisition is not blocked. If the permit is directly available, the provided
      * {@link ActionListener} will be called on the calling thread. During calls of
-     * {@link #blockOperations(long, TimeUnit, CheckedRunnable)}, permit acquisition can be delayed. The provided ActionListener will
-     * then be called using the provided executor once operations are no longer blocked.
+     * {@link #syncBlockOperations(long, TimeUnit, CheckedRunnable)}, permit acquisition can be delayed. The provided {@link ActionListener}
+     * will then be called using the provided executor once operations are no longer blocked.
      *
      * @param onAcquired      {@link ActionListener} that is invoked once acquisition is successful or failed
      * @param executorOnDelay executor to use for delayed call
      * @param forceExecution  whether the runnable should force its execution in case it gets rejected
      */
-    public void acquire(ActionListener<Releasable> onAcquired, String executorOnDelay, boolean forceExecution) {
+    public void acquire(final ActionListener<Releasable> onAcquired, final String executorOnDelay, final boolean forceExecution) {
         if (closed) {
             onAcquired.onFailure(new IndexShardClosedException(shardId));
             return;
         }
-        Releasable releasable;
+        final Releasable releasable;
         try {
             synchronized (this) {
                 releasable = tryAcquire();
                 if (releasable == null) {
-                    // blockOperations is executing, this operation will be retried by blockOperations once it finishes
+                    assert delayed;
```

Comment:
can we used delay to control the flow here? I think it be easier to read that to make `tryAcquire` check on delay. WDYT?

```
diff --git a/core/src/main/java/org/elasticsearch/index/shard/IndexShardOperationPermits.java b/core/src/main/java/org/elasticsearch/index/shard/IndexShardOperationPermits.java
index 7fc56a1..ebab08d 100644
--- a/core/src/main/java/org/elasticsearch/index/shard/IndexShardOperationPermits.java
+++ b/core/src/main/java/org/elasticsearch/index/shard/IndexShardOperationPermits.java
@@ -190,10 +190,7 @@ final class IndexShardOperationPermits implements Closeable {
         final Releasable releasable;
         try {
             synchronized (this) {
-                releasable = tryAcquire();
-                if (releasable == null) {
-                    assert delayed;
-                    // operations are delayed, this operation will be retried by doBlockOperations once the delay is remoked
+                if (delayed) {
                     if (delayedOperations == null) {
                         delayedOperations = new ArrayList<>();
                     }
@@ -205,6 +202,13 @@ final class IndexShardOperationPermits implements Closeable {
                     } else {
                         delayedOperations.add(new ContextPreservingActionListener<>(contextSupplier, onAcquired));
                     }
+                    releasable = null;
+                } else {
+                    releasable = tryAcquire();
+                    assert releasable != null;
+                }
+                if (releasable == null) {
+                    assert delayed;
                     return;
                 }
             }
@@ -212,7 +216,10 @@ final class IndexShardOperationPermits implements Closeable {
             onAcquired.onFailure(e);
             return;
         }
-        onAcquired.onResponse(releasable);
+        if (releasable != null) {
+            // if it's null operations are delayed, this operation will be retried by doBlockOperations once the delay is remoked
+            onAcquired.onResponse(releasable);
+        }
     }
 
     @Nullable private Releasable tryAcquire() throws InterruptedException {

```